integration/skydb.test.ts

```diff
@@ -240,4 +246,190 @@ describe(`SkyDB end to end integration tests for portal '${portal}'`, () => {
+  // setJSON. You have to use the same revision number that you had when you
+  // called getJSON.
+  describe("getJSON/setJSON data race regression integration tests", () => {
+    const jsonOld = { message: 1 };
+    const jsonNew = { message: 2 };
+
+    const delays = [0, 10, 100, 500];
+
+    const concurrentAccessError = "Concurrent access prevented in SkyDB";
+    const registryUpdateError = "Unable to update the registry";
+
+    it.each(delays)(
+      "should not get old data when getJSON is called after setJSON on a single client with a '%s' ms delay and getJSON doesn't fail",
+      async (delay) => {
+        const { publicKey, privateKey } = genKeyPairAndSeed();
+
+        // Set the data.
+        await client.db.setJSON(privateKey, dataKey, jsonOld);
+
+        // Try to invoke the data race.
+        let receivedJson;
+        try {
+          // Get the data while also calling setJSON.
+          const getJSONFn = async function () {
+            // Sleep.
+            await new Promise((r) => setTimeout(r, delay));
+            return await client.db.getJSON(publicKey, dataKey);
+          };
+          [{ data: receivedJson }] = await Promise.all([getJSONFn(), client.db.setJSON(privateKey, dataKey, jsonNew)]);
+        } catch (e) {
+          if ((e as Error).message.includes(concurrentAccessError)) {
+            // The data race condition has been prevented and we received the expected error. Return from test early.
+            return;
+          }
+
+          // Unexpected error, throw.
+          throw e;
+        }
+
+        // Data race did not occur, getJSON should have latest JSON.
+        expect(receivedJson).toEqual(jsonNew);
+      }
+    );
+
+    it.each(delays)(
+      "should not be able to use old data when getJSON is called after setJSON on two different clients with a '%s' ms delay",
+      async (delay) => {
+        // Create two new clients with a fresh revision cache.
+        const client1 = new SkynetClient(portal);
+        const client2 = new SkynetClient(portal);
+        const { publicKey, privateKey } = genKeyPairAndSeed();
+
+        // Set the initial data.
+        await client1.db.setJSON(privateKey, dataKey, jsonOld);
+
+        // Call getJSON and setJSON concurrently on different clients -- both
+        // should succeeed.
+
+        // Get the data while also calling setJSON.
+        const getJSONFn = async function () {
+          // Sleep.
+          await new Promise((r) => setTimeout(r, delay));
+          return await client2.db.getJSON(publicKey, dataKey);
+        };
+        const [{ data: receivedJson }] = await Promise.all([
+          getJSONFn(),
+          client1.db.setJSON(privateKey, dataKey, jsonNew),
+        ]);
+
+        // If we got old data from getJSON, make sure that that client is not
+        // able to update that JSON.
+
+        expect(receivedJson).not.toBeNull();
+        if (receivedJson === jsonNew) {
+          return;
```

Comment:
Same as above. Kind of tricky case and I don't think I've coded tests like this before, so I don't know what to do.

Reply:
I resolved the above conversation. My only point is that usually having a `return` statement in a unit test or integration test hides a possible bug. We could be under the impression that the assertions following this if condition keep passing, but in reality the test might return here 100% of the time.

If you've manually confirmed that is not the case, and we do run the assertions below in a decent amount of cases I guess it's fine. Just wanted to make the remark because I've experienced a situation like this before and since then a `return` in a unit test is a trigger for me.

Reply:
But you can obviously resolve the comment if you feel it's fine and harmless.

Reply:
To be fair it is an integration test and not a unit test, and there are two possible real-world conditions that could happen here that we want to explicitly test for. I appreciate that it's not ideal test design, but I don't know how else to write it.

> If you've manually confirmed that is not the case

I have, and will add a comment to that effect.